### PR TITLE
ONE: Green Sun's Twilight and support

### PR DESCRIPTION
**Pull request by Northmoc**
Some changes to DigMultipleEffect for this

### CODE REVIEW

File: forge-gui/res/cardsfolder/upcoming/green_suns_twilight.txt [hunk 1]
```diff
@@ -0,0 +1,12 @@
+Name:Green Sun's Twilight
+ManaCost:X G
+Types:Sorcery
+A:SP$ DigMultiple | DigNum$ SVar$X/Plus.1 | Reveal$ True | ChangeValid$ Creature,Land | ChangeLater$ True | Optional$ True | RememberChanged$ True | ImprintRest$ True | SubAbility$ DBBranch | SpellDescription$ Reveal the top X plus one cards of your library. Choose a creature card and/or a land card from among them. Put those cards into your hand and the rest on the bottom of your library in a random order. If X is 5 or more, instead put the chosen cards onto the battlefield or into your hand and the rest on the bottom of your library in a random order.
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/green_suns_twilight.txt:4]: I don't think it's Optional$ but min=1 & max=2?

File: forge-gui/res/cardsfolder/upcoming/green_suns_twilight.txt [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+SVar:DBBranch:DB$ Branch | BranchConditionSVar$ X | BranchConditionSVarCompare$ GE5 | TrueSubAbility$ DBChoose | FalseSubAbility$ Hand | SubAbility$ RestBottom
+SVar:DBChoose:DB$ GenericChoice | Choices$ Battlefield,Hand
+SVar:Battlefield:DB$ ChangeZone | Defined$ Remembered | Origin$ Library | Destination$ Battlefield | NoShuffle$ True | SubAbility$ RestBottom | SpellDescription$ Put the chosen card(s) onto the battlefield
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/green_suns_twilight.txt:7]: These other `SubAbility$ RestBottom` shouldn't be needed?

File: forge-gui/res/cardsfolder/upcoming/green_suns_twilight.txt (hunk 1 above)
Comment (Northmoc) [reply forge-gui/res/cardsfolder/upcoming/green_suns_twilight.txt:4]: Hmm, that would mean almost every DigMultiple card needs refactoring?

File: forge-gui/res/cardsfolder/upcoming/green_suns_twilight.txt (hunk 2 above)
Comment (Northmoc) [reply forge-gui/res/cardsfolder/upcoming/green_suns_twilight.txt:7]: Oops, yeah those are artifacts of old version :(

File: forge-gui/res/cardsfolder/upcoming/green_suns_twilight.txt (hunk 1 above)
Comment (tool4ever) [reply forge-gui/res/cardsfolder/upcoming/green_suns_twilight.txt:4]: How come? all others have "may"

Comment (Northmoc) [reply forge-gui/res/cardsfolder/upcoming/green_suns_twilight.txt]: I think I have this sorted...

File: forge-gui/res/cardsfolder/upcoming/green_suns_twilight.txt [hunk 3]
```diff
@@ -0,0 +1,12 @@
+Name:Green Sun's Twilight
+ManaCost:X G
+Types:Sorcery
+A:SP$ DigMultiple | DigNum$ SVar$X/Plus.1 | Reveal$ True | ChangeValid$ Creature,Land | ChangeLater$ True | RememberChanged$ True | ImprintRest$ True | SubAbility$ DBBranch | SpellDescription$ Reveal the top X plus one cards of your library. Choose a creature card and/or a land card from among them. Put those cards into your hand and the rest on the bottom of your library in a random order. If X is 5 or more, instead put the chosen cards onto the battlefield or into your hand and the rest on the bottom of your library in a random order.
+SVar:DBBranch:DB$ Branch | BranchConditionSVar$ X | BranchConditionSVarCompare$ GE5 | TrueSubAbility$ DBChoose | FalseSubAbility$ Hand | SubAbility$ RestBottom
+SVar:DBChoose:DB$ GenericChoice | Choices$ Battlefield,Hand
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/green_suns_twilight.txt:6]: Could avoid this line by using `DestinationAlternative$`

File: forge-game/src/main/java/forge/game/ability/effects/DigMultipleEffect.java [hunk 4]
```diff
@@ -79,6 +79,14 @@ public void resolve(SpellAbility sa) {
             } else {
                 CardCollection chosen = chooser.getController().chooseCardsForEffectMultiple(validMap, sa, Localizer.getInstance().getMessage("lblChooseCards"), chooseOptional);
```
Comment (tool4ever) [inline forge-game/src/main/java/forge/game/ability/effects/DigMultipleEffect.java:80]:
```
                CardCollection chosen;
                //ensure choosing something when possible and not optional
                while (true) {
                    chosen = chooser.getController().chooseCardsForEffectMultiple(validMap, sa,
                            Localizer.getInstance().getMessage("lblChooseCards"), chooseOptional);

                    if (!chosen.isEmpty()) {
                        game.getAction().reveal(chosen, chooser, true, Localizer.getInstance().getMessage("lblPlayerPickedCardFrom", chooser.getName()));
                        break;
                    }
                    if (chooseOptional) {
                        break;
                    }
                    chooser.getController().notifyOfValue(sa, null,
                            Localizer.getInstance().getMessage("lblMustChoose"));
                }
```

File: forge-gui/res/cardsfolder/upcoming/green_suns_twilight.txt (hunk 3 above)
Comment (Northmoc) [reply forge-gui/res/cardsfolder/upcoming/green_suns_twilight.txt:6]: Ha ha nice, my first thought was to change Illuna too... look who got to it 😉

**Comment by Hanmac** [comment]
we can update it later when the release notes are revealed

**Comment by Northmoc** [comment]
> we can update it later when the release notes are revealed

I will add it to my "check later" issue

**Review by tool4ever — APPROVED** [review]

**Pull request merged**